### ISSUE TITLE
[RFC]: Per-position target token logprobs for prefill scoring

### ISSUE THREAD

**Issue by aoshen02** [vllm-project/vllm#56860]
### Motivation.

**Status:** Draft

#### Summary

Add an inference/scoring-only interface that lets the caller specify a different set of target token IDs for each input position. The model returns only the logits or logprobs of those tokens at the corresponding positions.

The primary use case is Top-K scoring for On-Policy Distillation (OPD):

1. The student rollout records its own Top-K token IDs and logprobs at every response position.
2. The teacher runs a single forward pass over the full prompt plus the student response.
3. At each position, the teacher gathers only the candidate IDs supplied by the student and returns their logprobs (or logits).

The logical candidate tensor has shape `[batch, sequence_length, K]`. The number of candidates may differ between positions, so the implementation needs to support ragged lists.

#### Background

The existing `SamplingParams.logprob_token_ids` lets each request specify one fixed set of token IDs. This fits classification, label scoring, and similar workloads, where every scored row of a request uses the same candidates.

The OPD Top-K access pattern is different. The candidate sets of a single response may be:

```text
position 0: [A, B, C]
position 1: [B, D, E]
position 2: [A, F, G]
```

Taking the global union of the IDs across all positions and applying that union at every position gives correct results, but produces a large amount of unneeded data. With response length `R`, `K` candidates per position, and global union size `U`:

```text
data actually needed:              O(R × K)
global-union intermediate result:  O(R × U)
worst case:                        O(R² × K)
```

Each position should therefore carry only the candidate IDs it actually needs.

This RFC defines only the generic scoring capability; it does not prescribe how candidate sets are produced. Candidates may come from:

- student Top-K;
- teacher Top-K;
- the union/intersection of student and teacher Top-K;
- a caller-defined candidate set.

#### Goals

- Score arbitrary token IDs at each position during prompt/prefill scoring.
- Return only the IDs the caller requested and their values, without producing a full-vocabulary host/HTTP payload.
- Support raw logits and normalized logprobs, following the existing `logprobs_mode` semantics.
- Keep the existing request-wise `logprob_token_ids` behavior unchanged.
- Support continuous batching and ragged lists whose candidate count differs by position.
- Define causal alignment explicitly: the candidates at position `t` correspond to "predicting the token at position `t` from the prefix before position `t`".

#### Non-goals

- Implementing OPD or any distillation loss in vLLM.
- Deciding whether to use student Top-K or teacher Top-K.
- Changing token sampling or forcing the model to output a particular token.
- Returning full-vocabulary logits by default.

#### Proposed interface

Add an engine-level parameter, tentatively named `logprob_token_ids_positions`:

```python
SamplingParams(
    max_tokens=0,
    logprob_token_ids_positions=[
        [],          # nothing requested at this position
        [101, 202],  # candidate IDs for this position
        [7, 8, 9],
    ],
)
```

The exact wire name can be settled by the API maintainers. The key semantics are that every scored input position has its own set of candidate IDs. For multiple requests in a batch, each request has its own position-aligned list.

The result should keep both IDs and values, preserving the candidate order passed by the caller:

```text
position 0: ids=[],          values=[]
position 1: ids=[101, 202],  values=[...]
position 2: ids=[7, 8, 9],   values=[...]
```

The first causal position may keep an empty/placeholder entry for compatibility with the existing prompt-logprob output convention. Callers can restrict the returned range with `logprob_start_len`.

#### Execution model

1. The model runs one normal prompt/prefill forward over the supplied prompt and response token IDs.
2. The logits processor applies whatever normalization is configured.
3. A GPU gather reads

   ```python
   logits[position, candidate_ids[position]]
   ```

   or the corresponding logprobs.

4. Only the IDs and values of the ragged candidates are copied to the output.

For an OPD teacher request, prompt positions use empty lists, each response position carries the student's candidate IDs, and the teacher generates no new tokens (`max_tokens=0`).

Note: the teacher still has to run the forward over the full prompt+response. This interface only avoids returning token scores that are not needed; it does not compute hidden states for the response alone.

#### Compatibility and limitations

- The existing `logprob_token_ids` remains the fast path for "one fixed set of IDs per request".
- The initial implementation should support prompt/prefill scoring only; the decode path can be extended later if there is a clear need.
- Candidate IDs must be bounds-checked against the vocabulary.
- Behavior for duplicate IDs must be defined: either keep duplicates stably or reject them.
- The output must preserve the candidate order passed by the caller.
- HTTP/OpenAI-compatible exposure can be discussed separately later; nested lists add parameter-validation and serialization cost.

##### M1: Prefill request-wise fixed-ID scoring

First, support one fixed set of candidate IDs per request (the global union):

```python
prompt_logprob_token_ids=[101, 202, 303]
```

This is intended only for `max_tokens=0` prefill/scoring requests. The teacher runs the forward over the full input and gathers this set of IDs at every scored position. This stage reuses the existing request-wise token state and GPU gather path, and focuses on validating the output format, the logits/logprobs semantics, and position alignment under chunked prefill.

M1 is implemented in #54335.

##### M2: Prefill per-position sparse scoring

Add an interface with an independent candidate list per position:

```python
prompt_logprob_token_ids_positions=[
    [], [101, 202], [7, 8, 9],
]
```

The first version is still limited to prefill/scoring, `max_tokens=0`, and non-streaming requests. Decode-time per-position candidates, speculative/beam search, multimodal inputs, prefix cache, and the OpenAI-compatible endpoint are not supported yet. It supports multi-request batches, ragged candidate lists, and chunked prefill. Chunked prefill must carry each chunk's absolute start position so that logits rows stay aligned with candidate positions; otherwise the request must be rejected explicitly rather than silently returning misaligned results.

Internally, the first version can use a fixed-width `[num_rows, max_k]` candidate tensor plus a mask and restore ragged lists at output time; a fully packed sparse tensor is not required.

### Feedback Period.

At least one week.

#### Related implementations

Miles's OPD implementation has a path compatible with older SGLang: it first takes the global union of the candidate IDs across all positions, then queries through the request-wise `token_ids_logprob`. For example, when the per-position candidates are `[A, B]`, `[C, D]`, and `[A, E]`, it sends the fixed list `[A, B, C, D, E]`; the teacher returns the logprobs of these five tokens at every position, and Miles then filters out the entries each position actually needs. This path requires no server-side change but returns unneeded data, at `O(R × |U|)` scale (`U` being the union).

The optional `token_ids_logprob_positions` path instead sends a candidate list per position, and the teacher gathers only each position's own IDs, reducing the response from dense `O(R × |U|)` to sparse `O(R × K)`. Both paths compute the same result; they differ only in how candidates are transmitted and in the gather range.

SGLang's `top_logprobs_num` can already return the IDs and values of the natural Top-K at every position; its older `token_ids_logprob` is a request-wise fixed ID list. This RFC aims to support both "per-position candidate selection" and "gathering by specified IDs" in vLLM.

#### Open questions

- Should the parameter be named `logprob_token_ids_positions`, or something else?
- Should raw logits reuse the existing `logprobs_mode`, or should a separate scoring result be introduced?
- Should only the engine API be provided at first, without exposing an OpenAI-compatible endpoint?
- What is the most efficient representation of ragged candidates for CUDA graphs and TP/DP attention?
- Should position alignment use empty lists, or an explicit start offset / packed `(position, IDs)` pairs?

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
